libraries/tarball.rb

```diff
@@ -0,0 +1,157 @@
+      @options = options
+    end
+
+    attr_reader :prefix
+
+    def get_file(name)
+      data = []
+      target_path = @prefix.nil? ? name : "#{@prefix}/#{name}"
+      iterate file: proc { |path, entry| data << entry.read if path == target_path }, other: proc {}
+      fail "Multiple entries for #{target_path} found in #{@file_name}" if data.size > 1
+      data.first
+    end
+
+    def extract(extract_dir)
+      file_handler = proc do |path, entry|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path
+        File.open(target, 'wb') { |f| f.print entry.read }
+        FileUtils.chmod entry.header.mode, target
+        FileUtils.chown @options[:user], @options[:group], target
+      end
+
+      dir_handler = proc do |path, entry|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path, true
+        FileUtils.chmod entry.header.mode, target
+      end
+
+      symlink_handler = proc do |path, entry, other = {}|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path
+        File.unlink target if File.exist? target
+        File.symlink other[:linkname], target
+        FileUtils.chmod entry.header.mode, target
+        FileUtils.chown @options[:user], @options[:group], target
+      end
+
+      iterate file: file_handler, directory: dir_handler, symlink: symlink_handler
+    end
+
+    # Count the number of (qualified) entries in the tarball where the return value
+    # of the corresponding callback applied to each entry in the tarfile is truthy.
+    # With no arguments, counts the number of qualified entries
+    # If called with a block (only), the block is the callback to for every entry.
+    # If called with a map, invoke the appropriate callback for each corresponding entry.
+    # See iterate for details on how callbacks are invoked.
+    def count(callbacks = {}, &other)
+      callbacks = { other: other } if other && callbacks.empty?
+      i = 0
+      # Wrap each callback such that if it's truthy, we increment our counter.
+      hsh = callbacks.each_with_object({}) { |(k, v), h| h[k] = proc { |*args| i += 1 if v.call(args) } }
+      hsh[:other] = proc { i += 1 } if hsh.empty?
+      hsh[:other] = proc {} unless hsh[:other]
+      iterate hsh
+      i
+    end
+
+    def num_files
+      count file: proc { true }
+    end
+
+    def make_dirs(extract_dir, path, all = false)
+      segments = path.split('/')
+      segments = segments[0...-1] unless all
+      segments.each_with_object(extract_dir.dup) do |segment, target|
+        target << "/#{segment}"
+        FileUtils.mkdir target unless File.directory? target
+        FileUtils.chown @options[:user], @options[:group], target
+      end
+    end
+
+    # Iterate over the entries of the tarfile, and invoke the corresponding callback with each entry.
+    # We handle the LongFileName and LongSymLink entries as found (no callbacks, but they change values sent with other callbacks)
```

Comment:
First pass at some more doc on the tarball class. Does this help? bb7eb4c2e7c40f6457b9651d74e1280aab142a65